README.md
# Digger [![Build Status](https://travis-ci.org/CyberShadow/Digger.svg?branch=master)](https://travis-ci.org/CyberShadow/Digger) [![AppVeyor](https://ci.appveyor.com/api/projects/status/tm98i6iw931ma3yg/branch/master?svg=true)](https://ci.appveyor.com/project/CyberShadow/digger)

Digger can:

- build and test D from [git](https://github.com/dlang)
- build older versions of D
- build D plus forks and pending pull requests
- bisect D's history to find where regressions are introduced (or bugs fixed)

### Requirements

On POSIX, Digger needs git, g++, binutils and make. gcc-multilib and g++-multilib or equivalent are required for targeting x86 on x86_64 systems.

On Windows, Digger will download and unpack everything it needs (Git, DMC, DMD, 7-Zip, WiX, VS2013 and Windows SDK components).

### Get Digger

The easiest way to obtain and run Digger is through Dub. As Dub is included with DMD, simply run:

    $ dub fetch digger
    $ dub run digger -- ARGS...

where `ARGS...` are the Digger command and arguments (see below).

### Command-line usage

(The command line examples below assume that Digger is installed in a location under `PATH` -
replace `digger ...` with `./digger ...` or `dub run digger -- ...` as appropriate.)

##### Building D

    # build latest master branch commit
    $ digger build

    # build a specific D version
    $ digger build v2.064.2

    # build for x86-64
    $ digger build --model=64 v2.064.2

    # build commit from a point in time
    $ digger build "@ 3 weeks ago"

    # build latest 2.065 (release) branch commit
    $ digger build 2.065

    # build specified branch from a point in time
    $ digger build "2.065 @ 3 weeks ago"

    # build with added pull request
    $ digger build "master + dmd#123"

    # build with added GitHub fork branch
    $ digger build "master + Username/dmd/awesome-feature"

    # build with reverted commit
    $ digger build "master + -dmd/0123456789abcdef0123456789abcdef01234567"

    # build with reverted pull request
    $ digger build "master + -dmd#123"

##### Building D programs

    # Run the last built DMD version
    $ digger run - -- dmd --help

    # Build and run latest DMD master
    $ digger run master -- dmd --help

    # Build latest DMD master, and then build and run a D program using it
    $ digger run master -- dmd -i -run program.d

##### Hacking on D

    # check out git master (or some other version)
    $ digger checkout

    # build / incrementally rebuild current checkout
    $ digger rebuild

    # run tests
    $ digger test

Run `digger` with no arguments for detailed usage help.

##### Installing

Digger does not build all D components - only those that change frequently and depend on one another.
You can get a full package by upgrading an installation of a stable DMD release with Digger's build result:

    # upgrade the DMD in your PATH with Digger's result
    $ digger install

You can undo this at any time by running:

    $ digger uninstall

Successive installs will not clobber the backups created by the first `digger install` invocation,
so `digger uninstall` will revert to the state from before you first ran `digger install`.

You can also simultaneously install 32-bit and 64-bit versions of Phobos by first building and installing a 32-bit DMD,
then a 64-bit DMD (`--model=64`). `digger uninstall` will revert both actions.

To upgrade a system install of DMD on POSIX, simply run `digger install` as root, e.g. `sudo digger install`.

`digger install` should be compatible with [DVM](https://github.com/jacob-carlborg/dvm) or any other DMD installation.

Installation and uninstallation are designed with safety in mind.
When installing, Digger will provide detailed information and ask for confirmation before making any changes
(you can use the `--dry-run` / `--yes` switches to suppress the prompt).
Uninstallation will refuse to remove files if they were modified since they were installed,
to prevent accidentally clobbering user work (you can use `--force` to override this).

##### Bisecting

To bisect D's history to find which pull request introduced a bug, first copy `bisect.ini.sample` to `bisect.ini`, adjust as instructed by the comments, then run:

    $ digger bisect path/to/bisect.ini

If Digger ends up with a master/stable merge as the bisection result, switch the branches on the starting points accordingly, e.g.:

- If you specified `good=v2.080.0` and `bad=v2.081.0`, try `good=master@v2.080.0` and `bad=master@v2.081.0`
- If you specified `good=@2018-01-01` and `bad=@2019-01-01`, try `good=stable@2018-01-01` and `bad=stable@2019-01-01`
- Note that the master/stable branch-offs/merges do not happen at the same time as when releases are tagged,
  so you may need to increase the bisection range accordingly. See [DIP75](https://wiki.dlang.org/DIP75) for details.

### Configuration

You can optionally configure a few settings using a configuration file.
To do so, copy `digger.ini.sample` to `digger.ini` and adjust as instructed by the comments.

### Building

    $ git clone --recursive https://github.com/CyberShadow/Digger
    $ cd Digger
    $ rdmd --build-only digger

* If you get a link error, you may need to add `-allinst` or `-debug` due to [a DMD bug](https://github.com/CyberShadow/Digger/issues/37).

* A [dub](https://code.dlang.org/) definition is also included.

* On Windows, you may see:

      Warning 2: File Not Found version.lib

  This is a benign warning.

### Hacking

##### Backend

The code which builds D and manages the git repository is located in the [ae library](https://github.com/CyberShadow/ae)
(`ae.sys.d` package), so as to be reusable.

Currently, the bulk of the code is in [`ae.sys.d.manager`](https://github.com/CyberShadow/ae/blob/master/sys/d/manager.d).

`ae.sys.d.manager` clones [a meta-repository on BitBucket](https://bitbucket.org/cybershadow/d), which contains the major D components as submodules.
The meta-repository is created and maintained by another program, [D-dot-git](https://github.com/CyberShadow/D-dot-git).

The build requirements are fulfilled by the [`ae.sys.install`](https://github.com/CyberShadow/ae/tree/master/sys/install) package.

##### Frontend

Digger is the frontend to the above library code, implementing configuration, bisecting, etc.

Module list is as follows:

- `config` - configuration
- `repo` - customized D repository management, revision parsing
- `bisect` - history bisection
- `custom` - custom build management
- `install` - installation
- `digger` - entry point and command-line UI
- `common` - shared helpers

### Remarks

##### Wine

Digger should work fine under Wine. For 64-bit builds, you must first run `winetricks vcrun2013`.
Digger cannot do this automatically as this must be done from outside Wine.


CHANGELOG.md
Digger Changelog
================

Digger v3.0 (2020-04-21)
------------------------

 * Major internal changes for improved reliability
   * Cache version bumped to 3
 * Updated, backwards-incompatible .ini settings
   * All settings are now equally available from `digger.ini`, `bisect.ini` 
     and command line
   * Search for the configuration file according to the XDG Base Directory
     Specification
 * Add `-c` options to specify arbitrary `digger.ini` and `bisect.ini`
   settings on the command line
   * Specifying a `bisect.ini` for the `bisect` command is now optional,
     and can be entirely substituted with `-c` options.
 * Add `digger checkout` command, which simply checks out a given D revision 
   (`master` by default)
 * Add `digger test` command, to run tests for working tree state
 * Add `digger run` command, which runs arbitrary commands in the
   context of some D version
 * Add ability to revert a branch or pull request.  
   The syntax is to prefix the branch or PR with a `-` (minus sign).  
   Example: `digger build "master + -phobos#1234"`
 * Add ability to specify commit SHA1 instead of a branch or PR number.  
   Example: `digger build "master + -dmd/0123456789abcdef0123456789abcdef01234567"`
 * Add ability to specify a tag instead of a timestamp to use that tag's time
   Example: `digger build "master @ v2.085.0"`
 * Add `tools`, `extras` and `curl` components
 * Add `32mscoff` model support for Windows
 * Add `--jobs` option for controlling the GNU make `-j` parameter
 * Add `components.dmd.releaseDMD` build option to complement `debugDMD`
 * Add `components.dmd.dmdModel` option, which allows building a 64-bit
   `dmd.exe` on Windows (also supports `32mscoff`).
 * Improve bootstrapping up to arbitrary depths
 * Refuse to clobber working tree changes not done by Digger
 * Verify integrity of all downloaded files
 * Only download/install Visual Studio components as-needed
 * Prevent git from loading user/system configuration
 * Add `app.gitOptions` setting, which allows e.g. specifying an
   alternate way to clone git:// URLs
 * Add `dub.sdl`
 * Add `DIGGER_REVISION` environment variable during bisection
 * Add `extraSpec` bisect option, which allows specifying additional
   patches to apply during bisection
 * Add test suite
   * Enable continuous integration on Travis and AppVeyor
 * Remove `digger-web`
 * Various fixes

Digger v2.4 (2015-10-05)
------------------------

 * Fetch tags explicitly when updating
   (fixes some "unknown /ambiguous revision" errors)
 * Prepend result `bin` directory to `PATH`
   (fixes behavior when a `dmd` binary was installed in `/usr/bin`)
 * Add support for the `debugDMD` build option on POSIX
 * Fix incorrect repository tree order when using `git` cache engine
 * Fix `rebuild` ignoring build options on the command-line
 * Automatically install KindleGen locally when building website
 * Update OptLink installer
 * Download platform-specific DMD release packages
   (contributed by Martin Nowak)

Digger v2.3 (2015-06-14)
------------------------

 * Add `bisectBuild` bisect config option
 * Add `--with` and `--without` switches to control D components to build
 * Add `website` component for building dlang.org (POSIX-only)
 * Work around `appender` memory corruption bug with `git` cache engine
 * Various fixes

Digger v2.2 (2015-06-05)
------------------------

 * Fix `digger install` to work with `object.d`
 * Improve resilience of `digger install`
 * Add `--bootstrap` switch to build compiler entirely from C++ source
 * Replace usage of `git bisect run` with internal implementation
   * Bisection now prefers cached builds when choosing a commit to test
 * Allow customizing the set of components to build during bisection
 * Use git plumbing in git cache driver for concurrency and better performance
 * Don't cache build failures if the error is likely temporary

Digger v2.1 (2015-05-03)
------------------------

 * Add [license](LICENSE.md)
 * Add `git` cache engine
 * Add `cache` action and subcommands
 * Fix starting `digger-web` in OS X
   (auto-correct working directory)

Digger v2.0 (2015-04-26)
------------------------

 * `idgen.d` update (DMD now requires DMD to build)
 * Full core overhaul, for improved performance, granularity and extensibility.
   A fresh install is recommended.

Digger v1.1 (2015-03-04)
------------------------

 * Add `rebuild` action, for incremental rebuilds
   (thanks to Sergei Nosov)
 * Add `install` and `uninstall` actions
 * Add `--help` text
 * Add `--make-args` option
 * Add `--model` option to replace the `--64` switch
 * Add `--host` and `--port` to `digger-web`
 * Various smaller improvements

Digger v1.0 (2014-09-18)
------------------------

 * On Windows, Digger may now download and locally install (unpack) required 
   software, as needed:
   - Git
   - A number of Visual Studio 2013 Express and Windows SDK components (for 
     64-bit builds)
   - 7-Zip and WiX (necessary for unpacking Visual Studio Express components)
 * Various smaller improvements

Digger v0.3 (2014-05-22) [DConf edition]
----------------------------------------

 * Allow merging arbitrary GitHub forks
 * Add `--offline`, which suppresses updating the D repositories.
 * Move digger-web tasks to digger, thus removing D building logic from 
   digger-web binary
 * Improve revision parsing, allowing e.g. `digger build 2.065 @ 3 weeks ago`
 * Rename `digger-web` directory to `web`, to avoid conflict with POSIX binary 
   of `digger-web.d`
 * Fix web UI behavior when refreshing
 * Fix exit status code propagation
 * Various smaller improvements

Digger v0.2 (2014-04-01) [April Fools' edition]
-----------------------------------------------

 * Add `digger-web`
 * Fix parsing Environment configuration section
 * Various smaller improvements

Digger v0.1 (2014-02-17) [Initial release]
------------------------------------------

 * Initial announcement



digger.d
module digger;

import std.array;
import std.exception;
import std.file : thisExePath, exists;
import std.path;
import std.process;
import std.stdio;
import std.typetuple;

static if(!is(typeof({import ae.utils.text;}))) static assert(false, "ae library not found, did you clone with --recursive?"); else:

version (Windows)
	static import ae.sys.net.wininet;
else
	static import ae.sys.net.curl;

import ae.sys.d.manager : DManager;
import ae.utils.funopt;
import ae.utils.main;
import ae.utils.meta : structFun;
import ae.utils.text : eatLine;

import bisect;
import common;
import config;
import custom;
import install;
import repo;

// http://d.puremagic.com/issues/show_bug.cgi?id=7016
version(Windows) static import ae.sys.windows;

alias BuildOptions(string action, string pastAction, bool showBuildActions = true) = TypeTuple!(
	Switch!(hiddenOption, 0, "64"),
	Option!(string, showBuildActions ? "Select models (32, 64, or, on Windows, 32mscoff). You can specify multiple models by comma-separating them.\nOn this system, the default is " ~ DManager.Config.Build.components.common.defaultModel ~ " [build.components.common.models]" : hiddenOption, null, 0, "model"),
	Option!(string[], "Do not " ~ action ~ " a component (that would otherwise be " ~ pastAction ~ " by default). List of default components: " ~ DManager.defaultComponents.join(", ") ~ " [build.components.enable.COMPONENT=false]", "COMPONENT", 0, "without"),
	Option!(string[], "Specify an additional D component to " ~ action ~ ". List of available additional components: " ~ DManager.additionalComponents.join(", ") ~ " [build.components.enable.COMPONENT=true]", "COMPONENT", 0, "with"),
	Option!(string[], showBuildActions ? `Additional make parameters, e.g. "HOST_CC=g++48" [build.components.common.makeArgs]` : hiddenOption, "ARG", 0, "makeArgs"),
	Switch!(showBuildActions ? "Bootstrap the compiler (build from C++ source code) instead of downloading a pre-built binary package [build.components.dmd.bootstrap.fromSource]" : hiddenOption, 0, "bootstrap"),
	Switch!(hiddenOption, 0, "use-vc"),
	Switch!(hiddenOption, 0, "clobber-local-changes"),
);

enum specDescription = "D ref (branch / tag / point in time) to build, plus any additional forks or pull requests. Example:\n" ~
	"\"master @ 3 weeks ago + dmd#123 + You/dmd/awesome-feature\"";

void parseBuildOptions(T...)(T options) // T == BuildOptions!action
{
	if (options[0])
		d.config.build.components.common.models = ["64"];
	if (options[1])
		d.config.build.components.common.models = options[1].value.split(",");
	foreach (componentName; options[2])
		d.config.build.components.enable[componentName] = false;
	foreach (componentName; options[3])
		d.config.build.components.enable[componentName] = true;
	d.config.build.components.common.makeArgs ~= options[4];
	d.config.build.components.dmd.bootstrap.fromSource |= options[5];
	d.config.build.components.dmd.useVC |= options[6];
	d.verifyWorkTree = !options[7];
	static assert(options.length == 8);
}

struct Digger
{
static:
	@(`Build D from source code`)
	int build(BuildOptions!("build", "built") options, Parameter!(string, specDescription) spec = "master")
	{
		parseBuildOptions(options);
		buildCustom(spec);
		return 0;
	}

	@(`Incrementally rebuild the current D checkout`)
	int rebuild(BuildOptions!("rebuild", "rebuilt") options)
	{
		parseBuildOptions(options);
		incrementalBuild();
		return 0;
	}

	@(`Run tests for enabled components`)
	int test(BuildOptions!("test", "tested") options)
	{
		parseBuildOptions(options);
		runTests();
		return 0;
	}

	@(`Check out D source code from git`)
	int checkout(BuildOptions!("check out", "checked out", false) options, Parameter!(string, specDescription) spec = "master")
	{
		parseBuildOptions(options);
		.checkout(spec);
		return 0;
	}

	@(`Run a command using a D version`)
	int run(
		BuildOptions!("build", "built") options,
		Parameter!(string, specDescription ~ "\nSpecify \"-\" to use the previously-built version.") spec,
		Parameter!(string[], "Command to run and its arguments (use -- to pass switches)") command)
	{
		if (spec == "-")
			enforce(options == typeof(options).init, "Can't specify build options when using the last built version.");
		else
		{
			parseBuildOptions(options);
			buildCustom(spec, /*asNeeded*/true);
		}

		auto binPath = resultDir.buildPath("bin").absolutePath();
		environment["PATH"] = binPath ~ pathSeparator ~ environment["PATH"];

		version (Windows)
			return spawnProcess(command).wait();
		else
		{
			execvp(command[0], command);
			errnoEnforce(false, "execvp failed");
			assert(false); // unreachable
		}
	}

	@(`Install Digger's build result on top of an existing stable DMD installation`)
	int install(
		Switch!("Do not prompt", 'y') yes,
		Switch!("Only print what would be done", 'n') dryRun,
		Parameter!(string, "Directory to install to. Default is to find one in PATH.") installLocation = null,
	)
	{
		enforce(!yes || !dryRun, "--yes and --dry-run are mutually exclusive");
		.install.install(yes, dryRun, installLocation);
		return 0;
	}

	@(`Undo the "install" action`)
	int uninstall(
		Switch!("Only print what would be done", 'n') dryRun,
		Switch!("Do not verify files to be deleted; ignore errors") force,
		Parameter!(string, "Directory to uninstall from. Default is to search PATH.") installLocation = null,
	)
	{
		.uninstall(dryRun, force, installLocation);
		return 0;
	}

	@(`Bisect D history according to a bisect.ini file`)
	int bisect(
		Switch!("Skip sanity-check of the GOOD/BAD commits.") noVerify,
		Option!(string[], "Additional bisect configuration. Equivalent to bisect.ini settings.", "NAME=VALUE", 'c', "config") configLines,
		Parameter!(string, "Location of the bisect.ini file containing the bisection description.") bisectConfigFile = null,
	)
	{
		return doBisect(noVerify, bisectConfigFile, configLines);
	}

	@(`Cache maintenance actions (run with no arguments for details)`)
	int cache(string[] args)
	{
		static struct CacheActions
		{
		static:
			@(`Compact the cache`)
			int compact()
			{
				d.optimizeCache();
				return 0;
			}

			@(`Delete entries cached as unbuildable`)
			int purgeUnbuildable()
			{
				d.purgeUnbuildable();
				return 0;
			}

			@(`Migrate cached entries from one cache engine to another`)
			int migrate(string source, string target)
			{
				d.migrateCache(source, target);
				return 0;
			}
		}

		return funoptDispatch!CacheActions(["digger cache"] ~ args);
	}

	// hidden actions

	int buildAll(BuildOptions!("build", "built") options, string spec = "master")
	{
		parseBuildOptions(options);
		.buildAll(spec);
		return 0;
	}

	int delve(bool inBisect)
	{
		return doDelve(inBisect);
	}

	int parseRev(string rev)
	{
		stdout.writeln(.parseRev(rev));
		return 0;
	}

	int show(string revision)
	{
		d.getMetaRepo().git.run("log", "-n1", revision);
		d.getMetaRepo().git.run("log", "-n1", "--pretty=format:t=%ct", revision);
		return 0;
	}

	int getLatest()
	{
		writeln((cast(DManager.Website)d.getComponent("website")).getLatest());
		return 0;
	}

	int help()
	{
		throw new Exception("For help, run digger without any arguments.");
	}

	version (Windows)
	int getAllMSIs()
	{
		d.getVSInstaller().getAllMSIs();
		return 0;
	}
}

int digger()
{
	version (D_Coverage)
	{
		import core.runtime;
		dmd_coverSetMerge(true);
	}

	static void usageFun(string usage)
	{
		import std.algorithm, std.array, std.stdio, std.string;
		auto lines = usage.splitLines();

		stderr.writeln("Digger v" ~ diggerVersion ~ " - a D source code building and archaeology tool");
		stderr.writeln("Created by Vladimir Panteleev <vladimir@thecybershadow.net>");
		stderr.writeln("https://github.com/CyberShadow/Digger");
		stderr.writeln();
		stderr.writeln("Configuration file: ", opts.configFile.value.exists ? opts.configFile.value : "(not present)");
		stderr.writeln("Working directory: ", config.config.local.workDir);
		stderr.writeln();

		if (lines[0].canFind("ACTION [ACTION-ARGUMENTS]"))
		{
			lines =
				[lines[0].replace(" ACTION ", " [OPTION]... ACTION ")] ~
				getUsageFormatString!(structFun!Opts).splitLines()[1..$] ~
				lines[1..$];

			stderr.writefln("%-(%s\n%)", lines);
			stderr.writeln();
			stderr.writeln("For help on a specific action, run: digger ACTION --help");
			stderr.writeln("For more information, see README.md.");
			stderr.writeln();
		}
		else
			stderr.writefln("%-(%s\n%)", lines);
	}

	return funoptDispatch!(Digger, FunOptConfig.init, usageFun)([thisExePath] ~ (opts.action ? [opts.action.value] ~ opts.actionArguments : []));
}

mixin main!digger;
